[RFC] Make Output Ranges Using Methods The Preferred toString Option

JackStouffer opened the pull request:
This proposal

* Modifies `std.format` to recognize any `toString` that accepts a character accepting output range as having a valid `toString` overload
* Modifies `std.format` to prioritize and use the output range method when formatting aggregate types if it exists
* De-emphasizes the `scope delegate` `toString` methods in the docs

## Rationale

Firstly, speed. [According to Walter](https://issues.dlang.org/show_bug.cgi?id=9785#c3), the delegate passed as a parameter can never be in-lined because of how DMD does data flow analysis. Using instead the defined `put` method of an output range would solve that problem.

Secondly, the vision of Phobos.  Ranges are the present and the future. Output ranges really need more support in Phobos (std.conv).

Thirdly, safety. The vast majority of the code I've seen doesn't mark the `delegate` as `@safe` to allow non-safe sinks to work with it. This means that the `toString` method itself can never be marked `@safe` either. Making `toString` a template who's safety relies on the `put` method of the output range side-steps this problem entirely.

JackStouffer commented:
Ho man, this is a fun bug. Turns out, if you don't define your `FormatSpec` as `const ref` in the parameters of your `toString` (and why would you when the docs don't tell you to do that), it wasn't recognized as a valid overload by `formatObject` before this change.

Case in point, BigInt's `toString(scope void delegate(const(char)[]) sink, ref FormatSpec!char f)` overload wasn't ever called by std.format. It called the string format option every time.

Fix incoming.

JackStouffer commented:
https://github.com/dlang/phobos/pull/5992

quickfur commented on std/format.d:
```diff
@@ -1735,10 +1735,22 @@ FormatSpec!Char singleSpec(Char)(Char[] fmt)
+ * ---
+ *
+ * Where `W` allows a character accepting
+ * $(REF_ALTTEXT ouput range, isOutputRange, std,_range,primitives) with its template constraints.
```
Nitpick: ouput -> output

quickfur commented on std/format.d:
```diff
@@ -1735,10 +1735,22 @@ FormatSpec!Char singleSpec(Char)(Char[] fmt)
+ * string toString();
+ * ---
+ *
+ * Where `W` allows a character accepting
```
What about "Where W is an output range that accepts ...".

jmdavis replied:
Also, it really should be `ref W`. Accepting an output range by value is begging for trouble.

quickfur commented on std/format.d:
```diff
@@ -1735,10 +1735,22 @@ FormatSpec!Char singleSpec(Char)(Char[] fmt)
+ * The template type does not have to be called `W`.
+ *
+ * The following overloads are also accepted for legacy reasons. It's recommended that
+ * any new code forgo these overloads for speed and attribute acceptance reasons.
```
What about also accepting a `toString` overload that takes an output range that can accept any range of characters?  If efficiency is a concern here, it would make sense for `toString` to be able to take a, output range of char/wchar/dchar instead of needing a function call roundtrip per character.

jmdavis replied:
Is it really just for legacy reasons though? What about classes? As soon as you templatize `toString`, it can't be virtual.

quickfur replied:
Hmm, that's true. Virtual `toString` methods could be problematic because `format` may not know how to construct an output range of a type that will be accepted by `toString`.

Perhaps one way out is use `std.range.interfaces.OutputRangeObject`. Since such an object is an output range, it can be recognized by `format`, and since it gives an opaque interface to arbitrary underlying output range implementations, it will also give us the flexibility to choose the best implementation used by `format`.

OTOH, since `scope void delegate(const(char)[])` is already established in current code, and already allows arbitrary implementations, perhaps we should just retain support for it.  Code that can afford to have `toString` templatized can still take advantage of a more flexible implementation of `format`.

quickfur replied:
`@safe`ty is still an issue for virtual `toString` functions, though. Well, as well as other attributes.  I've no idea how to work around that.

quickfur commented:
In general I like the direction this is going.  The main concern I have is that we should support `toString` taking an output range that accepts ranges of characters, in order to minimize roundtrips when outputting strings.

JackStouffer commented:
Changed the requirements to look for a `string` accepting output range instead of a `char` accepting one. Also added checks to make sure the output range is being taken by ref.

As for virtualizing `toString`, IMO the best option would be to use the `string toString()` overload in that case, as it's probably going to be faster with the inlining issue I mentioned. Plus, it can also be `@safe` in that case.

JackStouffer commented:
https://github.com/dlang/phobos/pull/5995

JackStouffer commented:
Thought about the round-trip per character issue some more, and I remembered that `put` solves this problem automatically for the user. Emphasized it in the docs.

quickfur commented:
@JackStouffer It took me a few minutes to understand why and how exactly `std.range.put` solves the problem of round-trip per character.  You might want to put in a slightly more detailed explanation than just pointing the reader to `put`. :-) Also, it would be nice if there's a hyperlink to the documentation for `put` somewhere.

JackStouffer commented:
@quickfur I'm not planning on deprecating the other overloads anytime soon, if ever. There's just too much code out there that uses it, and this is such a common function, that I'm not comfortable removing the delegate overloads from `std.format`.

quickfur commented on std/format.d:
```diff
@@ -1735,10 +1735,21 @@ FormatSpec!Char singleSpec(Char)(Char[] fmt)
+ * which accepts characters. The template type does not have to be called `W`.
+ *
+ * The following overloads are also accepted for legacy reasons. It's recommended that
+ * any new code forgo these overloads for speed and attribute acceptance reasons.
```
You might want to reword this, if the older overloads are still recommended for class hierarchies where `toString` needs to be virtual.

JackStouffer commented:
@quickfur Addressed comments

quickfur commented on std/format.d:
```diff
@@ -3484,11 +3553,59 @@ template hasToString(T, Char)
+    }
+    // not using the overload enum to not break badly defined toString overloads
+    // e.g. defining the FormatSpec as ref and not const ref led this function
+    // to ignore that toString overload
```
Just a thought: to prevent similar breakages in the future, what about if the new unittest added above is changed to have non-trivial return values for the `toString` overloads, and the unittest actually checks for these values?  The idea being that if a particular `toString` overload is ignored or not called correctly for whatever reason, the unittest would catch it.

wilzbach commented on std/format.d:
```diff
@@ -1917,13 +1929,53 @@ void formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, const ref F
+
+        void toString(W)(ref W writer, const ref FormatSpec!char f) if (isOutputRange!(W, char))
+        {
+            // make sure to use std.range.primitives.put and not writer.put directly
```
Mentioning this in the summary page.

wilzbach commented on std/format.d:
```diff
@@ -1917,13 +1929,53 @@ void formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, const ref F
+ * this time using the `scope delegate` method.
  *
- *  `formatValue` also allows to reuse existing format specifiers:
+ * This method is now discouraged. Please use the output range method
```
`$(RED ..)` the first sentence. Or maybe just bold?

wilzbach commented on std/format.d:
```diff
@@ -3455,13 +3507,30 @@ if (is(AssocArrayTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
+        {T val = void; FormatSpec!Char f; static struct S {void put(Char s){}}
+        S s; val.toString(s, f);
+        // force toString to take output range by ref
+        static assert(ParameterStorageClassTuple!(T.toString!(S))[0] == ParameterStorageClass.ref_);}
```
How about enforcing that the second parameter (FormatSpec) is const and ref?

JackStouffer replied:
If it's ref compilation will already fail if it's not const

wilzbach commented on std/format.d:
```diff
@@ -1735,10 +1735,22 @@ FormatSpec!Char singleSpec(Char)(Char[] fmt)
- * const void toString(scope void delegate(const(char)[]) sink);
- * const string toString();
+ * void toString(W)(ref W w, const ref FormatSpec fmt)
+ * void toString(W)(ref W w)
```
You don't have a public unittest showing this to the user. I know it's trivial, but I still would add one to avoid guaranteed questions on the NG.

wilzbach commented on std/format.d:
```diff
@@ -1917,13 +1929,50 @@ void formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, const ref F
+
+    formatValue(w, p, spec);
+    assert(w.data == "(16,11)");
+}
```
This public tests is failing, run with:

```
make -f posix.mak std/format.publictests
```

wilzbach replied:
The line number on the error message is the one from the generated file.
(it probably makes sense to insert empty lines in it, s.t. the line numbers match up)

JackStouffer replied:
I'm not really sure how to debug this. This is green on the autotester

wilzbach replied:
1) Compile Phobos
2) Run the extracted test with Phobos-dev:

```
unittest
{
    import std.array : appender;
    import std.format;

    static struct Point
    {
        int x, y;

        void toString(W)(ref W writer, const ref FormatSpec!char f) if (isOutputRange!(W, char))
        {
            // std.range.primitives.put
            put(writer, "(");
            formatValue(writer, x, f);
            put(writer, ",");
            formatValue(writer, y, f);
            put(writer, ")");
        }
    }

    auto w = appender!string();
    auto spec = singleSpec("%s");
    auto p = Point(16, 11);

    formatValue(w, p, spec);
    assert(w.data == "(16,11)");
}
```

BTW seeing that we just catched this issue, you might want to add more public tests (or tests in a different module).

wilzbach replied:
The issue is that `w.data` is `Point(16, 11)` here.

wilzbach commented on std/format.d:
```diff
@@ -1917,13 +1929,50 @@ void formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, const ref F
+    {
+        int x, y;
+
+        void toString(W)(ref W writer, const ref FormatSpec!char f) if (isOutputRange!(W, char))
```
If constraints should be in the line of their declaration:

```
void toString(W)(ref W writer, const ref FormatSpec!char f)
if (isOutputRange!(W, char))
```

JackStouffer commented:
@wilzbach Fixed. The public tests didn't have access to `std.range.primitives`.

JackStouffer commented:
Ping @andralex

schveiguy commented on changelog/toString.dd:
```diff
@@ -0,0 +1,30 @@
+writeln(t); // writes "Custom toString"
+-------
+
+This has several benefits for the user. Firstly, this design is much friendlier
```
Firstly -> First, Secondly -> Second...

schveiguy commented:
This is cool, I like it.

CyberShadow commented:
This pull request seems to have exposed an `-allinst` DMD bug, thus causing a regression:
https://issues.dlang.org/show_bug.cgi?id=19406
